storage.py

```diff
@@ -0,0 +1,235 @@
+
+from clcache import CacheFileStrategy, getStringHash, printTraceStatement, CompilerArtifacts, \
+    CACHE_COMPILER_OUTPUT_STORAGE_CODEC
+
+
+class CacheDummyLock(object):
+    def __enter__(self):
+        pass
+
+    def __exit__(self, typ, value, traceback):
+        pass
+
+
+class CacheMemcacheStrategy(object):
+    def __init__(self, server, cacheDirectory=None, manifestPrefix='manifests_', objectPrefix='objects_'):
+        self.fileStrategy = CacheFileStrategy(cacheDirectory=cacheDirectory)
+        # XX Memcache Strategy should be independent
+
+        self.lock = CacheDummyLock()
+        self.localCache = {}
+        self.localManifest = {}
+        self.objectPrefix = objectPrefix
+        self.manifestPrefix = manifestPrefix
+
+        self.connect(server)
+
+    def connect(self, server):
+        server = CacheMemcacheStrategy.splitHosts(server)
+        assert server, "{} is not a suitable server".format(server)
+        if len(server) == 1:
+            clientClass = Client
+            server = server[0]
+        else:
+            from pymemcache.client.hash import HashClient
+            clientClass = HashClient
+            server = server
+        self.client = clientClass(server, ignore_exc=True,
+                                  serializer=python_memcache_serializer,
+                                  deserializer=python_memcache_deserializer,
+                                  timeout=5,
+                                  connect_timeout=5,
+                                  key_prefix=(getStringHash(self.fileStrategy.dir) + "_").encode("UTF-8")
+                                 )
+        # XX key_prefix ties fileStrategy cache to memcache entry
+        # because tests currently the integration tests use this to start with clean cache
+        # Prevents from having cache hits in when code base is in different locations
+        # adding code to production just for testing purposes
+
+    def server(self):
+        return self.client.server
+
+    @staticmethod
+    def splitHost(host):
+        port = 11211
+        index = host.rfind(':')
+        if index != -1:
+            host, port = host[:index], int(host[index + 1:])
+        if not host or port > 65535:
+            raise ValueError
+        return host.strip(), port
+
+    @staticmethod
+    def splitHosts(hosts):
+        """
+        :param hosts: A string in the format of HOST:PORT[,HOST:PORT]
+        :return: a list [(HOST, int(PORT)), ..] of tuples that can be consumed by socket.connect()
+        """
+        return [CacheMemcacheStrategy.splitHost(h) for h in hosts.split(',')]
+
+    def __str__(self):
+        return "Remote Memcache @{} object-prefix: {}".format(self.server, self.objectPrefix)
+
+    # XX Global lock here?
+    # XX memcache has own statistics about hit/misses
+    @property
+    def statistics(self):
+        return self.fileStrategy.statistics
+
+    # XX Global lock here?
```

Comment:
I think my comment is valid: `with cache.statistics.lock, cache.statistics as stats:` in `ensureArtifactsExists` is a global barrier. Where MemcachedStrategy could suffer local locking. It is fine if there are enough parallel compilations though